--- a/src/mktplinkfw2.c
+++ b/src/mktplinkfw2.c
@@ -150,6 +150,12 @@ static struct flash_layout layouts[] = {
 		.kernel_ep	= 0x80000000,
 		.rootfs_ofs	= 0x140000,
 	}, {
+                .id             = "16Mmtk",
+                .fw_max_len     = 0xfd0000,
+                .kernel_la      = 0x80000000,
+                .kernel_ep      = 0x80000000,
+                .rootfs_ofs     = 0x140000,
+        }, {
 		.id		= "8MSUmtk", /* Split U-Boot OS */
 		.fw_max_len	= 0x770000,
 		.kernel_la	= 0x80000000,
